gobblin-runtime/src/main/java/gobblin/runtime/JobContext.java

```diff
@@ -350,77 +351,105 @@ void commit() throws IOException {
     }
 
     for (Map.Entry<String, JobState.DatasetState> entry : this.datasetStatesByUrns.get().entrySet()) {
-      String datasetUrn = entry.getKey();
-      JobState.DatasetState datasetState = entry.getValue();
-      finalizeDatasetStateBeforeCommit(datasetState);
-
-      Class<? extends DataPublisher> dataPublisherClass;
-      try (Closer closer = Closer.create()) {
-        dataPublisherClass = getJobDataPublisherClass(datasetState)
-            .or((Class<? extends DataPublisher>) Class.forName(ConfigurationKeys.DEFAULT_DATA_PUBLISHER_TYPE));
-        if (!canCommitDataset(datasetState)) {
-          this.logger.warn(String.format("Not committing dataset %s of job %s with commit policy %s and state %s",
-              datasetUrn, this.jobId, this.jobCommitPolicy, datasetState.getState()));
-          allDatasetsCommit = false;
-          if (UnpublishedHandling.class.isAssignableFrom(dataPublisherClass)) {
-            DataPublisher publisher = closer.register(DataPublisher.getInstance(dataPublisherClass, datasetState));
-            this.logger
-                .info(String.format("Calling publisher to handle unpublished work units for dataset %s of job %s.",
-                    datasetUrn, this.jobId));
-            ((UnpublishedHandling) publisher).handleUnpublishedWorkUnits(datasetState.getTaskStatesAsWorkUnitStates());
-          }
-          continue;
+      // NOTE: order here is important as we don't want to skip processDatasetCommit if there was
```

Comment:
If you use `&` then order doesn't matter. If use `&&` then order matters.
In fact you can simply use `&=`